src/core/algorithms/dc/verifier/dc_verifier.cpp

```diff
@@ -0,0 +1,521 @@
+
+DCVerifier::DCVerifier() : Algorithm({}) {
+    using namespace config::names;
+
+    RegisterOptions();
+    MakeOptionsAvailable({kTable});
+};
+
+void DCVerifier::RegisterOptions() {
+    DESBORDANTE_OPTION_USING;
+
+    RegisterOption(Option<std::string>(&dc_string_, kDenialConstraint, kDDenialConstraint, ""));
+    RegisterOption(config::kTableOpt(&input_table_));
+}
+
+void DCVerifier::MakeExecuteOptsAvailable() {
+    MakeOptionsAvailable({config::names::kDenialConstraint});
+}
+
+void DCVerifier::LoadDataInternal() {
+    data_ = model::CreateTypedColumnData(*input_table_, true);
+    input_table_->Reset();
+    relation_ = ColumnLayoutRelationData::CreateFrom(*input_table_, true);
+}
+
+unsigned long long int DCVerifier::ExecuteInternal() {
+    auto start = std::chrono::system_clock::now();
+    dc::DC dc;
+    try {
+        dc = SplitDC(dc_string_);
+    } catch (std::exception const& e) {
+        LOG(INFO) << e.what();
+        return 0;
+    }
+
+    bool has_header = !relation_->GetSchema()->GetColumns().front().get()->GetName().empty();
+    index_offset_ = 1 + static_cast<size_t>(has_header);
+    result_ = Verify(dc);
+
+    auto elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::system_clock::now() - start);
+
+    return elapsed_time.count();
+}
+
+bool DCVerifier::Verify(dc::DC const& dc) {
+    dc::DC temp_dc = ConvertEqualities(dc);
+    std::vector<dc::Predicate> no_diseq_preds, diseq_preds, predicates = temp_dc.GetPredicates();
+
+    for (auto& pred : predicates) {
+        auto op = pred.GetOperator();
+        if (op.GetType() == dc::OperatorType::kUnequal and pred.IsCrossTuple()) {
+            diseq_preds.emplace_back(pred);
+        } else {
+            no_diseq_preds.emplace_back(pred);
+        }
+    }
+
+    size_t cur_dc_num = 0;
+    size_t preds_count = diseq_preds.size();
+    size_t all_comb_count = std::pow(2, preds_count);
+    dc::DCType dc_type = GetType(GetDC(no_diseq_preds, diseq_preds, cur_dc_num));
+    auto check = kDCTypeToVerificationMethod.at(dc_type);
+
+    // TODO: check the article for optimization 2^l -> 2^(l-1) dc's
+
+    // Consider 'cur_signs' as a set of operators where each digit
+    // in binary representation (0 or 1) means '<' or '>' respectively
+    // e.g. 21 = 10101 is ">, <, >, <, >"
+    for (size_t cur_signs = 0; cur_signs < all_comb_count; ++cur_signs) {
+        dc::DC cur_dc = GetDC(no_diseq_preds, diseq_preds, cur_signs);
+        if (!(this->*check)(cur_dc)) return false;
+    }
+
+    return true;
+}
+
+dc::DC DCVerifier::GetDC(std::vector<dc::Predicate> const& no_diseq_preds,
+                         std::vector<dc::Predicate> const& diseq_preds, size_t cur_signs) {
+    size_t preds_count = diseq_preds.size();
+    std::vector<dc::Operator> operators = {dc::Operator("<"), dc::Operator(">")};
+    std::vector<dc::Predicate> cur_predicatates(no_diseq_preds);
+    cur_predicatates.reserve(preds_count);
```

Comment:
So, why don't just
```
std::vector<dc::Predicate> cur_predicatates(no_diseq_preds);
cur_predicatates.reserve(diseq_preds_count + no_diseq_preds.size());
```
?

Reply:
A vector owns a memory region of some size (this is capacity). When this region is resized, like after a `reserve` call, it acquires a new memory region and moves the previous elements into it, one-by-one.
Your way would first allocate a memory chunk to copy `no_diseq_preds` into, then `reserve` would allocate a new memory chunk, to where the copied elements would be moved. Reserving first avoids that: it allocates the chunk first, then copies the elements.
`reserve` doesn't increase the size of the memory chunk, it tries to set it: if the vector has capacity 5 and `reserve(2)` is called, it won't increase its capacity to 7 (or greater), it will try to (in a typical implementation) set it to 2, see that `2 < 5`, and refuse (do nothing).

Reply:
That's fair, haven't considered case when the size is bigger then requested reserve size